File: api/py/[slug].js
'use strict';

const cp   = require('child_process');
const fs   = require('fs');
const os   = require('os');
const path = require('path');
const { initializeApp, getApps } = require('firebase/app');
const { getDatabase, ref, get, set } = require('firebase/database');

// ── Firebase ──────────────────────────────────────────────────────────
const firebaseConfig = {
  apiKey     : "AIzaSyBLzocO-22pXwjNg7HFG75OQzaFjbfV_0k",
  authDomain : "watch-be6e0.firebaseapp.com",
  databaseURL: "https://watch-be6e0-default-rtdb.firebaseio.com",
  projectId  : "watch-be6e0",
  storageBucket    : "watch-be6e0.firebasestorage.app",
  messagingSenderId: "1044218513896",
  appId      : "1:1044218513896:web:14db4db1291a4903dcbc5d",
};
const firebaseApp = getApps().length===0 ? initializeApp(firebaseConfig) : getApps()[0];
const db = getDatabase(firebaseApp);

// ── Helpers ───────────────────────────────────────────────────────────
function decodeKey(k){
  return k
    .replace(/__dot__/g,'.').replace(/__hash__/g,'#')
    .replace(/__dollar__/g,'$').replace(/__slash__/g,'/')
    .replace(/__lb__/g,'[').replace(/__rb__/g,']');
}

function page404(slug){
  return `<!DOCTYPE html><html><head><title>404 — Dropify</title>
<meta name="viewport" content="width=device-width,initial-scale=1">
<style>*{margin:0;padding:0;box-sizing:border-box}body{font-family:'Segoe UI',sans-serif;background:#f8fafc;display:flex;align-items:center;justify-content:center;min-height:100vh}
.box{text-align:center;max-width:400px;padding:40px}h1{font-size:64px;font-weight:800;color:#0d1117}p{color:#57606a;margin-top:8px}code{background:#f1f5f9;padding:2px 8px;border-radius:5px;font-size:13px}
a{display:inline-block;margin-top:20px;background:#3b82f6;color:#fff;text-decoration:none;border-radius:9px;padding:10px 22px;font-size:14px;font-weight:600}</style>
</head><body><div class="box"><h1>404</h1><p>Python app <code>${slug}</code> not found.</p><a href="/">Home</a></div></body></html>`;
}

function pageError(err){
  return `<!DOCTYPE html><html><head><title>Error</title>
<meta name="viewport" content="width=device-width,initial-scale=1">
<style>*{margin:0;padding:0;box-sizing:border-box}body{font-family:'Segoe UI',sans-serif;background:#0d1117;color:#e6edf3;display:flex;align-items:center;justify-content:center;min-height:100vh}
.box{max-width:640px;padding:32px;width:100%}h2{color:#f85149;margin-bottom:12px}pre{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:16px;font-size:12px;line-height:1.6;overflow-x:auto;white-space:pre-wrap;word-break:break-word}
a{display:inline-block;margin-top:14px;color:#58a6ff;font-size:13px}</style>
</head><body><div class="box"><h2>⚠ Python Runtime Error</h2><pre>${String(err).replace(/</g,'&lt;').replace(/>/g,'&gt;')}</pre><a href="javascript:history.back()">← Back</a></div></body></html>`;
}

// ── Main Handler ──────────────────────────────────────────────────────
module.exports = async function handler(req, res) {
  const slug = req.query.slug;
  if(!slug) return res.status(404).send(page404('unknown'));

  try{
    // Lookup slug
    const slugSnap = await get(ref(db, `pySlug/${slug}`));
    if(!slugSnap.exists()) return res.status(404).send(page404(slug));
    const { uid, appId } = slugSnap.val();

    // Get app
    const appSnap = await get(ref(db, `pythonApps/${uid}/${appId}`));
    if(!appSnap.exists()) return res.status(404).send(page404(slug));
    const app = appSnap.val();

    if(app.status !== 'active'){
      return res.status(503).send(`<!DOCTYPE html><html><body style="font-family:sans-serif;text-align:center;padding:60px;color:#555"><h2>🔒 App Offline</h2><p style="margin-top:8px">This app is currently inactive.</p></body></html>`);
    }

    const files = app.files  || {};
    const env   = app.envVars || {};
    const entry = app.entryFile ? decodeKey(app.entryFile) : 'main.py';

    // Create temp dir
    const tmpDir = path.join(os.tmpdir(), `pyapp_${Date.now()}_${Math.random().toString(36).slice(2)}`);
    fs.mkdirSync(tmpDir, { recursive: true });

    // Write all files
    for(const [key, content] of Object.entries(files)){
      const fname = decodeKey(key);
      const fpath = path.join(tmpDir, fname);
      fs.mkdirSync(path.dirname(fpath), { recursive: true });
      fs.writeFileSync(fpath, content || '');
    }

    // Request info available in Python
    const reqInfo = JSON.stringify({
      method : req.method,
      path   : req.url,
      query  : req.query,
      body   : req.body || {},
      headers: { 'content-type': req.headers['content-type']||'', 'user-agent': req.headers['user-agent']||'' },
    });

    // Wrapper: inject env, request, override print, run entry
    const entryCode = (files[app.entryFile] || app.code || '').replace(/\\/g,'\\\\').replace(/"""/g,'\\"\\"\\"');
    const wrapper = `import sys, os, json as _j
sys.path.insert(0,'${tmpDir.replace(/\\/g,'/')}')
_env=_j.loads("""${JSON.stringify(env).replace(/"""/g,'\\"\\"\\"')}""")
for k,v in _env.items(): os.environ[k]=str(v)
request=_j.loads("""${reqInfo.replace(/"""/g,'\\"\\"\\"')}""")
env=_env
_buf=[]
import builtins as _bt
_orig_print=_bt.print
def print(*a,sep=' ',end='\\n',**kw): _buf.append(sep.join(str(x) for x in a)+end)
_bt.print=print
try:
    exec(open('${tmpDir.replace(/\\/g,'/')}/${entry}').read())
except SystemExit: pass
except Exception as e: _buf.append(str(e))
_bt.print=_orig_print
sys.stdout.write(''.join(_buf))
`;

    const wrapperPath = path.join(tmpDir, '_run.py');
    fs.writeFileSync(wrapperPath, wrapper);

    // Install requirements.txt if present
    const reqTxt = path.join(tmpDir, 'requirements.txt');
    const pkgDir = path.join(tmpDir, '_pkgs');

    const runIt = () => new Promise(resolve => {
      const env2 = { ...process.env, PYTHONPATH: pkgDir+':'+tmpDir };
      cp.execFile('python3', [wrapperPath], { timeout: 8000, env: env2 },
        (err, stdout, stderr) => {
          try{ fs.rmSync(tmpDir,{recursive:true,force:true}); }catch(_){}
          resolve({ out: stdout||'', err: stderr||'', crash: err });
        }
      );
    });

    let result;
    if(fs.existsSync(reqTxt)){
      fs.mkdirSync(pkgDir,{recursive:true});
      await new Promise(r => cp.execFile('pip3',['install','-r',reqTxt,'--target',pkgDir,'-q','--no-deps'],{timeout:20000},()=>r()));
      result = await runIt();
    } else {
      result = await runIt();
    }

    // Increment views
    try{
      const vSnap = await get(ref(db,`pythonApps/${uid}/${appId}/views`));
      await set(ref(db,`pythonApps/${uid}/${appId}/views`),(vSnap.val()||0)+1);
    }catch(_){}

    // Error with no output
    if(result.crash && !result.out){
      return res.status(500).send(pageError(result.err || result.crash.message));
    }
    if(result.err && !result.out){
      return res.status(500).send(pageError(result.err));
    }

    const out = result.out.trim();

    // Auto-detect content type
    if(out.toLowerCase().startsWith('<!doctype') || out.startsWith('<html') || /<[a-z][\s\S]*>/i.test(out)){
      res.setHeader('Content-Type','text/html; charset=utf-8');
    } else {
      try{ JSON.parse(out); res.setHeader('Content-Type','application/json'); }
      catch(_){ res.setHeader('Content-Type','text/plain; charset=utf-8'); }
    }

    return res.status(200).send(out || '(no output — add print() to your code)');

  }catch(err){
    console.error('[PyHost]',err.message);
    return res.status(500).send(pageError(err.message));
  }
};

File: api/py/[slug].py
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import urllib.request
import urllib.error
import json
import sys
import io

FIREBASE_URL = "https://watch-be6e0-default-rtdb.firebaseio.com"

def fb_get(path):
    try:
        url = f"{FIREBASE_URL}/{path}.json"
        with urllib.request.urlopen(url, timeout=10) as r:
            return json.loads(r.read())
    except:
        return None

def fb_set(path, value):
    try:
        data = json.dumps(value).encode()
        req = urllib.request.Request(f"{FIREBASE_URL}/{path}.json", data=data, method='PUT')
        urllib.request.urlopen(req, timeout=5)
    except:
        pass

def decode_key(k):
    return (k.replace('__dot__','.').replace('__hash__','#')
             .replace('__dollar__','$').replace('__slash__','/')
             .replace('__lb__','[').replace('__rb__',']'))

def html_404(slug):
    return f"""<!DOCTYPE html><html><head><title>404 — Dropify</title>
<style>*{{margin:0;padding:0;box-sizing:border-box}}
body{{font-family:'Segoe UI',sans-serif;background:#f8fafc;display:flex;align-items:center;justify-content:center;min-height:100vh}}
.b{{text-align:center;padding:40px}}.b h1{{font-size:64px;font-weight:800;color:#111}}
.b p{{color:#666;margin-top:8px}}.b a{{display:inline-block;margin-top:16px;background:#3b82f6;
color:#fff;text-decoration:none;border-radius:9px;padding:10px 22px;font-size:14px;font-weight:600}}</style>
</head><body><div class="b"><h1>404</h1><p>Python app <code>{slug}</code> not found.</p>
<a href="/">← Home</a></div></body></html>"""

def html_error(err):
    safe = str(err).replace('<','&lt;').replace('>','&gt;')
    return f"""<!DOCTYPE html><html><head><title>Error</title>
<style>*{{margin:0;padding:0;box-sizing:border-box}}
body{{font-family:'Segoe UI',sans-serif;background:#0d1117;color:#e6edf3;
display:flex;align-items:center;justify-content:center;min-height:100vh;padding:20px}}
.b{{max-width:640px;width:100%}} h2{{color:#f85149;margin-bottom:12px;font-size:18px}}
pre{{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:16px;
font-size:12px;line-height:1.6;overflow-x:auto;white-space:pre-wrap;word-break:break-word}}
a{{color:#58a6ff;font-size:13px;display:block;margin-top:12px}}</style>
</head><body><div class="b"><h2>⚠ Python Runtime Error</h2>
<pre>{safe}</pre><a href="javascript:history.back()">← Back</a></div></body></html>"""


class handler(BaseHTTPRequestHandler):

    def do_GET(self):  self.handle_req()
    def do_POST(self): self.handle_req()

    def handle_req(self):
        parsed = urlparse(self.path)
        params = parse_qs(parsed.query, keep_blank_values=True)
        slug   = params.get('slug', [''])[0]

        if not slug:
            self._send(404, html_404('unknown'), 'text/html')
            return

        # Firebase lookup
        slug_data = fb_get(f"pySlug/{slug}")
        if not slug_data:
            self._send(404, html_404(slug), 'text/html')
            return

        uid    = slug_data.get('uid', '')
        app_id = slug_data.get('appId', '')
        app    = fb_get(f"pythonApps/{uid}/{app_id}")

        if not app:
            self._send(404, html_404(slug), 'text/html')
            return

        if app.get('status') != 'active':
            self._send(503,
                '<html><body style="font-family:sans-serif;text-align:center;padding:60px"><h2>🔒 App Offline</h2></body></html>',
                'text/html')
            return

        code     = app.get('code', '')
        env_vars = app.get('envVars') or {}

        # Query params (exclude 'slug')
        query = {k: (v[0] if len(v)==1 else v) for k, v in params.items() if k != 'slug'}

        # POST body
        body = {}
        if self.command == 'POST':
            try:
                length = int(self.headers.get('Content-Length', 0))
                if length:
                    raw = self.rfile.read(length)
                    try: body = json.loads(raw)
                    except: body = raw.decode('utf-8', errors='replace')
            except: pass

        request_obj = {
            'method' : self.command,
            'path'   : parsed.path,
            'query'  : query,
            'body'   : body,
            'headers': dict(self.headers),
        }

        # Capture stdout
        buf = io.StringIO()
        old_stdout = sys.stdout
        sys.stdout = buf

        try:
            globs = {
                '__builtins__' : __builtins__,
                'request'      : request_obj,
                'env'          : env_vars,
                'urllib'       : urllib,
                'json'         : json,
                'sys'          : sys,
                'io'           : io,
            }
            # Try to import requests if available
            try:
                import requests as _req
                globs['requests'] = _req
            except ImportError:
                pass

            exec(compile(code, '<bot>', 'exec'), globs)

        except SystemExit:
            pass
        except Exception as e:
            import traceback
            tb = traceback.format_exc()
            buf.write(html_error(tb))
        finally:
            sys.stdout = old_stdout

        output = buf.getvalue()

        # Increment views (async best-effort)
        try:
            old_views = fb_get(f"pythonApps/{uid}/{app_id}/views") or 0
            fb_set(f"pythonApps/{uid}/{app_id}/views", old_views + 1)
        except: pass

        if not output:
            output = '<html><body style="font-family:sans-serif;padding:40px;color:#555"><h3>No output</h3><p>Add <code>print()</code> to your code.</p></body></html>'

        # Detect content type
        stripped = output.strip()
        low = stripped[:100].lower()
        if low.startswith('<!doctype') or low.startswith('<html') or ('<' in low and '>' in low):
            ctype = 'text/html; charset=utf-8'
        else:
            try:
                json.loads(stripped)
                ctype = 'application/json'
            except:
                ctype = 'text/plain; charset=utf-8'

        self._send(200, output, ctype)

    def _send(self, code, body, ctype='text/html'):
        enc = body.encode('utf-8')
        self.send_response(code)
        self.send_header('Content-Type', ctype)
        self.send_header('Content-Length', str(len(enc)))
        self.end_headers()
        self.wfile.write(enc)

    def log_message(self, *a): pass  # suppress logs
